migrations/2022-10-06-183632_move_blocklist_to_db/up.sql

```diff
@@ -0,0 +1,177 @@
+
+-- Create allowlist and blocklist tables
+create table allowlist (
+  id serial primary key,
+  instance_id int references instance on update cascade on delete cascade not null unique,
+  published timestamp not null default now(),
+  updated timestamp null
+);
+
+create table blocklist (
+  id serial primary key,
+  instance_id int references instance on update cascade on delete cascade not null unique,
+  published timestamp not null default now(),
+  updated timestamp null
+);
+
+-- Move all the extra site settings-type columns to a local_site table
+-- Add a lot of other fields currently in the lemmy.hjson
+create table local_site (
+  id serial primary key,
+  site_id int references site on update cascade on delete cascade not null unique,
+
+  -- Site table fields
+  site_setup boolean default false not null,
+  enable_downvotes boolean default true not null,
+  open_registration boolean default true not null,
+  enable_nsfw boolean default true not null,
+  community_creation_admin_only boolean default false not null,
+  require_email_verification boolean default false not null,
+  require_application boolean default true not null,
+  application_question text default 'to verify that you are human, please explain why you want to create an account on this site'::text,
+  private_instance boolean default false not null,
+  default_theme text default 'browser'::text not null,
+  default_post_listing_type text default 'local'::text not null,
+  legal_information text,
+  hide_modlog_mod_names boolean default true not null,
+  application_email_admins boolean default false not null,
+
+  -- Fields from lemmy.hjson
+  slur_filter_regex text,
+  actor_name_max_length int default 20 not null,
+  rate_limit_message int default 180 not null,
+  rate_limit_message_per_second int default 60 not null,
+  rate_limit_post int default 6 not null,
+  rate_limit_post_per_second int default 600 not null,
+  rate_limit_register int default 3 not null,
+  rate_limit_register_per_second int default 3600 not null,
+  rate_limit_image int default 6 not null,
+  rate_limit_image_per_second int default 3600 not null,
+  rate_limit_comment int default 6 not null,
+  rate_limit_comment_per_second int default 600 not null,
+  rate_limit_search int default 60 not null,
+  rate_limit_search_per_second int default 600 not null,
+  federation_enabled boolean default true not null,
+  federation_debug boolean default false not null,
+  federation_strict_allowlist boolean default true not null,
+  federation_http_fetch_retry_limit int default 25 not null,
+  federation_worker_count int default 64 not null,
+  email_enabled boolean default false not null,
+  email_smtp_server varchar(255),
+  email_smtp_login varchar(255),
+  email_smtp_password varchar(255),
+  email_smtp_from_address varchar(255),
+  email_tls_type varchar(255) default 'none' not null,
+  captcha_enabled boolean default false not null,
+  captcha_difficulty varchar(255) default 'medium' not null,
```

Comment:
Another thing is that these settings are all documented at least a little bit with comments in config.hjson. If we move them into db, those comments will be gone and there wont be any docs at all about what they do.

Reply:
> I dont think it makes sense to move all of these into db. At least things like federation_http_fetch_retry_limit, worker count or email config should stay in config, so that a random site admin cant change them and break functionality.

Benefits of moving to DB:

- Live-updates
- Admins can change them instead of system administrators who have to know ssh.

My thinking is that we should move everything except for **wiring**, into the database. So obviously database settings, pictrs, opentelemetry url, site setup, and port binding are the only things that should really remain there, because they are necessary to wire lemmy to the database, its picture server, etc.

- Federation settings - Would be great to be able to turn on and off on the fly, change worker counts, etc. An admin should be able to turn off federation, without having to ask a system admin to do that for them. If we move some federation settings, we might as well move them all.
- Rate limits - Definitely should be in the DB.
- Email - Should be in database, because its great for admins to tweak that on the fly, but I could be convinced to move it back, if there's a strong case that admins should never have to change it. Could be considered "wiring".

IMO we shouldn't base the decision on powers, after all none of the settings are destructive. Admins can already do things like purge items (which is much more permanent than tweaking a setting).

> Another thing is that these settings are all documented at least a little bit with comments in config.hjson. If we move them into db, those comments will be gone and there wont be any docs at all about what they do.

This is a case where it actually gets *more* documented when we move them to the database. Remember that these will now become site setting UI fields, and will benefit from multi-language support. We can add as many descriptive strings for the site settings as we need.

> Also this seems to be the only reason which makes 64 columns feature necessary, and i dont think its worth it. At least split it into a separate table (local_site_config) to get rid of the feature.

The local_site table is the config, just like the local_user table is their site-specific config. 

Although it doesn't make logical sense to split that table up (you should really only add SQL tables when you need one-many relations), I could be convinced on maybe separating the rate_limits into their own table, if the compile time increase from that feature is a lot. I'll let you decide if that's worth it. Let me know.

Reply:
I suppose it makes sense, although personally i would definitely consider email as wiring. Consider the case of some managed hosting provider who preconfigures a working email relay. Dont want the costumer to mess up the config.

> This is a case where it actually gets more documented when we move them to the database. Remember that these will now become site setting UI fields, and will benefit from multi-language support. We can add as many descriptive strings for the site settings as we need.

Multi-language support is good, but its not documentation. Right now there is no documentation at all in lemmy-ui for existing site settings. If you want to move these, at least save the existing doc comments by putting them in [this page](https://join-lemmy.org/docs/en/administration/configuration.html). 

> Although it doesn't make logical sense to split that table up (you should really only add SQL tables when you need one-many relations), I could be convinced on maybe separating the rate_limits into their own table, if the compile time increase from that feature is a lot. I'll let you decide if that's worth it. Let me know.

It definitely increases a lot, you can try it for yourself by running `RUSTC_WRAPPER='' cargo build --timings` with and without the feature enabled.

Reply:
K, this is updated. I moved email back to the hjson, moved rate_limits to its own table, and renamed the allowlist tables.